### PR TITLE
Improve MAVLink behavior with flow control capable links

### PR DESCRIPTION
**Pull request by MUSTARDTIGERFPV**
With these changes, INAV can now be used with ExpressLRS MAVLink, including both RC & telemetry all in one!

You can watch a demo here: https://www.youtube.com/watch?v=d55q5sUbIkc

This _should_ be solid for both ELRS and other MAVLink telemetry radios, though I haven't tested with them yet. Wanted to get the PR created so folks can take a look at it. :)

### CODE REVIEW

File: src/main/telemetry/mavlink.c [hunk 1]
```diff
@@ -162,14 +162,16 @@ static serialPortConfig_t *portConfig;
 
 static bool mavlinkTelemetryEnabled =  false;
 static portSharing_e mavlinkPortSharing;
+static uint8_t txbuff_free = 100;
+static bool txbuff_valid = false;
 
 /* MAVLink datastream rates in Hz */
 static uint8_t mavRates[] = {
     [MAV_DATA_STREAM_EXTENDED_STATUS] = 2,      // 2Hz
-    [MAV_DATA_STREAM_RC_CHANNELS] = 5,          // 5Hz
```
Comment (mmosca) [inline src/main/telemetry/mavlink.c:169]: I have to check if we can make these rates configurable.

I know mavlinnk can be used to control some onboard peripherals, like gimbals. I don't know if they rely on RC_CHANNELS stream or if they have their own messages.

I am not sure if we should change default behavior, but we can certainly document recommended lower rates for telemetry radios with low bandwidth.

File: src/main/telemetry/mavlink.c [hunk 2]
```diff
@@ -1176,11 +1202,17 @@ static bool processMAVLinkIncomingTelemetry(void)
                 case MAVLINK_MSG_ID_MISSION_REQUEST:
                     return handleIncoming_MISSION_REQUEST();
                 case MAVLINK_MSG_ID_RC_CHANNELS_OVERRIDE:
-                    return handleIncoming_RC_CHANNELS_OVERRIDE();
+                    handleIncoming_RC_CHANNELS_OVERRIDE();
```
Comment (mmosca) [inline src/main/telemetry/mavlink.c:1205]: Not sure how I feel about this returning false. I will make some more comments where you are checking if you are good to transmit.

File: src/main/telemetry/mavlink.c [hunk 3]
```diff
@@ -1202,24 +1232,23 @@ void handleMAVLinkTelemetry(timeUs_t currentTimeUs)
...
-        if (!incomingRequestServed ||
-            (
-                 (rxConfig()->receiverType == RX_TYPE_SERIAL) &&
-                 (rxConfig()->serialrx_provider == SERIALRX_MAVLINK) &&
-                 !tristateWithDefaultOnIsActive(rxConfig()->halfDuplex)
-            )
-        ) {
-            processMAVLinkTelemetry(currentTimeUs);
-        }
+    if (shouldSendTelemetry) {
```
Comment (mmosca) [inline src/main/telemetry/mavlink.c:1254]: I think the wait before sending makes sense for half duplex links. You want to make sure the link is not busy when you start to transmit. A short wait to ensure the is silent is a common strategy for this. ELRS mavlink rc is delegating that de-conflicting down the line, so we can treat it as full duplex with flow control as we won't have collisions on the uart itself.

So something along the lines of:

if(!isHalfDuplex || (!incommingRequestServed)) would keep the current half duplex behavior, but we need to tell INAV to treat mavlinkrc as full duplex.

The check should probably be expanded to check for serialrx_halfduplex and telemetry_halfduplex.

I have to dig in the code a big more, and validate who owns the uart. The tristateWithDefaultOnIsActive  check on rxConfig is probably wrong, if telemetry_halfduplex is set to OFF. A workaround would be to set both to full duplex for validation.

File: src/main/telemetry/mavlink.c (hunk 1 above)
Comment (mmosca) [reply src/main/telemetry/mavlink.c:169]: Oh wait, they already are: 
``` 
# get mavlink 
mavlink_ext_status_rate = 2
Allowed range: 0 - 255

mavlink_rc_chan_rate = 5
Allowed range: 0 - 255

mavlink_pos_rate = 2
Allowed range: 0 - 255

mavlink_extra1_rate = 10
Allowed range: 0 - 255

mavlink_extra2_rate = 2
Allowed range: 0 - 255

mavlink_extra3_rate = 1
Allowed range: 0 - 255
```

File: src/main/telemetry/mavlink.c (hunk 3 above)
Comment (mmosca) [reply src/main/telemetry/mavlink.c:1254]: This check probably makes more sense for half duplex on vs off.
```
static bool isMAVLinkTelemetryHalfDuplex(void) {
    return  (rxConfig()->receiverType == RX_TYPE_SERIAL && rxConfig()->serialrx_provider == SERIALRX_MAVLINK && tristateWithDefaultOffIsActive(rxConfig()->halfDuplex))
            || telemetryConfig()->halfDuplex;
}
```

File: src/main/telemetry/mavlink.c (hunk 1 above)
Comment (MUSTARDTIGERFPV) [reply src/main/telemetry/mavlink.c:169]: I'm modifying the defaults in settings.yaml instead

**Comment by mmosca** [comment]
Did some basic testing and appears to be working. Will test a bit more and verify telemetry info is getting through.

**Comment by mmosca** [comment]
Mention change of default values for the mavlink message rates on release notes and that mavlink radio_status messages will ignore half duplex frame pacing and rely on txbuffer.

Transparent radio links will still behave as they did before, by default.

**Comment by sugatovo** [comment]
The other day I tested expresslrs mavlink on the inav and I want to share my observations. I want to say right away that I have just begun to work with GCS apps, mavlink, autonomous stuff and personal observations are all that I can provide at the moment. Now to the point.

elrs mavlink and ainav 8.0.0. Since I have a laptop from Apple, I am using QgroundControl as the GCS app. Connection between plane and QGC did not cause any problems, in this regard, everything went smoothly. The application itself perfectly tracks the the aircraft and its telemetry, however, when trying to load mission into the aircraft, an error constantly occurs. Various missions such as the Corridor Scan or Survey did not give any results and could not be flashed to FC, so QGC was used only for monitoring params and position. Next, I connected plane to mission planner and everything was more successful. Connection also did not cause any problems, it was enough to follow instructions from this link: https://www.expresslrs.org/software/mavlink/. Mission Planner made it possible to load mission into the flight controller and everything went well. However, you can't download anything other than a waypoint mission. When trying to loiter a point or land plane, Inav performs a return to home and does not want to do programmed mission (even though it was successfully uploaded to the plane) . At the moment, that's it. I will continue to try to figure out how it all works and do updates if I’ll come up with smth.

**Pull request merged**